### PR TITLE
Fix for multiple subscription requests.

### PR DESCRIPTION
**Pull request by noobs2ninjas**
The issue I found is that multiple subscription request in quick succession would cause the socket to throw an error.  This primarily happens because when the first subscription calls reconnect the socket is no longer nil. When the second socket calls subscribe the function will skip to  `sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))`

The error is thrown because the socket isnt nil...yet has not been completely connected. The simple fix is creating and setting a boolean that is set during the initial check...and isnt set back to false till the reconnect has finished.

I also fixed a little logic concerning disconnected. In a few places it was a little confusing and in disconnect even contradicted itself.

### CODE REVIEW

**Comment by noobs2ninjas** [comment]
Contributor License Agreement completed.

**Comment by nlutsenko** [comment]
Hey @noobs2ninjas, this looks promising.
The intention of `reconnect` is that you don't need to call it explicitly - it will be called on subscribing to any subscription, and we are always trying to stay connected and alive unless you want to explicitly `disconnect()`, which is what that boolean is being used for.

When the client becomes connected - it will try to subscribe to all of the subscriptions that were asked for, meaning that you don't need to handle any operation manually...

I believe we can do better than adding another variable - by reusing the existing `readyState` on a socket. Going to place a suggested place for it on the diff.

Let us know your thoughts, and whether the proposed solution works. If yes - please update the pull request to that change and we'll happily merge it!

File: Sources/ParseLiveQuery/Client.swift [hunk 1]
```diff
@@ -149,11 +152,11 @@ extension Client {
             )
             subscriptions.append(subscriptionRecord)
 
-            if socket == nil {
-                if !disconnected {
-                    reconnect()
-                }
-            } else {
+            if socket == nil && disconnected{
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: This could be simplified into this, and probably will be able to get rid of another boolean variable at all.

``` Swift
if socket?.readyState == .OPEN {
  sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))
} else if !disconnected && socket?.readyState != .CONNECTING {
  reconnect()
}
```

**Comment by noobs2ninjas** [comment]
I actually wasn't aware of the ready state! After looking at it I see that we have states .CLOSED which could replace disconnected(assuming you didn't have another intended purpose for it) , .CONNECTING which could replace reconnecting, and .OPEN.

I figured we could do this 

```
if socket == nil || socket?.readyState == .CLOSED{
            reconnect()
        } else if socket?.readyState == .OPEN{
            sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))
        }
```

We also still should be checking for nil as well correct?

**Comment by nlutsenko** [comment]
Unless we can guarantee the fact that the state never change s to CLOSED outside of being triggered by the user -  I don't think it would work and provide the same guarantees. :)
Also, this is minor, but you don't need to check for `socket == nil`, since optional chained unwrapping is going to do it for you.

**Comment by noobs2ninjas** [comment]
I wasnt thinking of it in the unwrapping. I was thinking that if the socket was nil...it cant have the ready state of .CLOSED but we still need to reconnect.

**Comment by nlutsenko** [comment]
Another piece, that I've missed - please revert all of the whitespace changes.

**Comment by nlutsenko** [comment]
Right, I somehow am still reading this in a context of previous code, sorry about that :(
What do you think in regards to still using a separate boolean variable for `disconnected()`?

**Comment by noobs2ninjas** [comment]
So, the biggest question is...where would we be checking disconnected? Im not opposed to it.  I just don't see where we could check it where it would be beneficial.

Even if they explicitly called close, they must want it back open, otherwise why would they be calling subscribe?

I can kinda see how we might check for that in the delegate call backs with errors, but Im almost feeling that we probably shouldn't be seeing that upon a user disconnect as we are setting everything to nil. Does the delegate stay open even if the socket is nil?

On the same subject as the delegate. Is there a default delay for reconnect upon initial fail? I seemed to have seen it the other night when I was playing around but couldn't find it.

**Comment by nlutsenko** [comment]
> So, the biggest question is...where would we be checking disconnected? Im not opposed to it. I just don't see where we could check it where it would be beneficial.
> Even if they explicitly called close, they must want it back open, otherwise why would they be calling subscribe?
> I can kinda see how we might check for that in the delegate call backs with errors, but Im almost feeling that we probably shouldn't be seeing that upon a user disconnect as we are setting everything to nil. Does the delegate stay open even if the socket is nil?

The portion I am talking about is explicit call to `client.disconnect()`.
Other than people explicitly do this - we always reconnect, but the moment when they disconnect - we aren't going to reconnect without a call to it.

**Comment by noobs2ninjas** [comment]
oh! you are talking about for use inside the SRWebSocketDelegate? gotcha!

**Comment by noobs2ninjas** [comment]
ok I simply changed disconnected to userDisconnected...so its a little more clear about what its specifically for. When I first saw disconnected I thought it was a simple boolean made to represent the sockets specific state.

File: Sources/ParseLiveQuery/Internal/ClientPrivate.swift [hunk 2]
```diff
@@ -129,21 +129,22 @@ extension Client: SRWebSocketDelegate {
     public func webSocketDidOpen(webSocket: SRWebSocket!) {
         // TODO: Add support for session token and user authetication.
         self.sendOperationAsync(.Connect(applicationId: applicationId, sessionToken: ""))
+
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift]: Nit: Please revert this change.

File: Sources/ParseLiveQuery/Client.swift [hunk 3]
```diff
@@ -149,13 +150,13 @@ extension Client {
...
             subscriptions.append(subscriptionRecord)
 
-            if socket == nil {
-                if !disconnected {
-                    reconnect()
-                }
-            } else {
-                sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))
-            }
+        if socket == nil || socket?.readyState == .CLOSED{
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Nit: Add space before `{`

File: Sources/ParseLiveQuery/Client.swift [hunk 4]
```diff
@@ -149,13 +150,13 @@ extension Client {
...
+            userDisconnected = false
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: I am not sure this is the right place for it.
This would trigger the reconnection on any subscription, whilst the user never explicitly called `reconnect()`.
Maybe adding a private method for `reconnect()` and moving this state change to that method would be better.

File: Sources/ParseLiveQuery/Client.swift (hunk 3 above)
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Also, I believe you don't need to check for `.CLOSED`, but rather to check for `!= .OPEN`, since if you are in the process of closing the socket - it won't reopen it, whilst it should.

**Comment by nlutsenko** [comment]
Few minor nits and a question for discussion. Awesome job so far!

File: Sources/ParseLiveQuery/Client.swift (hunk 4 above)
Comment (noobs2ninjas) [inline Sources/ParseLiveQuery/Client.swift]: Thats a good idea. Although I think to do this right, at some point we should either create a socket class(that perhaps sub-classes SRWebSocket) or a struct just for the socket itself. Then we would have a lot more options about what is private or public, and allow the socket to manage itself. This would also put us in the position to let the the primary delegate private(make ClientPrivate actually private) and let us create our own public delegate that removes all the SRWebSocketDelegate functions from public view.

For now lets simply to add a check for it. Speaking of which. I have a quick question involving explicit disconnect and then adding a resubscribe. I was thinking of printing 

> Warning: The client was explicitly disconnected! You must explicitly call .reconnect() in order to process your subscriptions.

Also, it probably wouldn't hurt to add to the documentation that a explicit reconnect requires a explicit reconnect.

Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: All of this is a great idea!
Feel like we can do the last 2 even in this pull request or pretty easily?

File: Sources/ParseLiveQuery/Client.swift [hunk 5]
```diff
@@ -203,7 +210,7 @@ extension Client {
             socket.delegate = self
             socket.setDelegateDispatchQueue(queue)
             socket.open()
-
+            userDisconnected = true
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Was this intended? Not sure I follow the idea here.

File: Sources/ParseLiveQuery/Internal/ClientPrivate.swift [hunk 6]
```diff
@@ -133,17 +133,17 @@ extension Client: SRWebSocketDelegate {
...
+        //set disconnect and try reconnecting
+        if !userDisconnected{
             reconnect()
         }
     }
 
     public func webSocket(webSocket: SRWebSocket!, didCloseWithCode code: Int, reason: String!, wasClean: Bool) {
         print("code: \(code) reason: \(reason)")
-
-        // TODO: Better retry logic, unless `disconnect()` was explicitly called
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift:145]: I feel like we should revert this comment, since this is a valid `todo` to improve the retry logic (exponential backoff and stuff).

File: Sources/ParseLiveQuery/Client.swift [hunk 7]
```diff
@@ -149,13 +150,18 @@ extension Client {
             )
             subscriptions.append(subscriptionRecord)
 
-            if socket == nil {
-                if !disconnected {
-                    reconnect()
-                }
-            } else {
-                sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))
+        if socket == nil || socket?.readyState != .OPEN{
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Looks like you don't need the second check and you could just reverse this control statement, meaning you can check whether the socket is open and send the subscribe event, otherwise check for not disconnected and stuff.
Would simplify the checking logic here.

Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Also, we might as well not do anything if the socket is in `CONNECTING` state, since we are going to destroy and re-create the socket, but we don't have to since the subscriptions were not sent yet.

File: Sources/ParseLiveQuery/Client.swift [hunk 8]
```diff
@@ -149,13 +150,18 @@ extension Client {
...
+            
+            if !userDisconnected{
+                reconnect()
+            }else{
+                print("Warning: The client was explicitly disconnected! You must explicitly call .reconnect() in order to process your subscriptions.")
             }
+            
+        } else if socket?.readyState == .OPEN{
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Nit: Space before `{`. I know this is tedious, but this is the overall code style that we are using.

**Comment by nlutsenko** [comment]
Few more nits and thoughts, almost there!

**Comment by noobs2ninjas** [comment]
Done. I also realized we still need to check for .CONNECTING otherwise we will run reconnect again

**Comment by noobs2ninjas** [comment]
Ok so I kinda get how you want whitespace to work now. HAHA sorry this took me so long to get. Never really worked with/for a company that ever had policies like that so its kind of the first time I've ever dealt with things like that.

**Comment by nlutsenko** [comment]
Whitespace changes are still there :) Please revert them.

**Comment by nlutsenko** [comment]
Also, if that's not a hassle - could you go through the changed code and make sure that you have a single space before `{` in every control statement?

**Comment by noobs2ninjas** [comment]
added spaces

File: Sources/ParseLiveQuery/Client.swift [hunk 9]
```diff
@@ -116,83 +116,86 @@ extension Client {
...
+        subscriptions.append(subscriptionRecord)
+        
+        if socket?.readyState == .OPEN {
+            sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))
+        } else if socket == nil || socket?.readyState != .CONNECTING {
+            if !userDisconnected {
+                reconnect()
             } else {
-                sendOperationAsync(.Subscribe(requestId: subscriptionRecord.requestId, query: query))
+                print("Warning: The client was explicitly disconnected! You must explicitly call .reconnect() in order to process your subscriptions.")
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Client.swift]: Maybe `debugPrint`? Not sure if this should be logged in production...

File: Sources/ParseLiveQuery/Internal/ClientPrivate.swift [hunk 10]
```diff
@@ -209,7 +209,7 @@ extension Client {
             let jsonEncoded = operation.JSONObjectRepresentation
             let jsonData = try NSJSONSerialization.dataWithJSONObject(jsonEncoded, options: NSJSONWritingOptions(rawValue: 0))
             let jsonString = String(data: jsonData, encoding: NSUTF8StringEncoding)
-
+
```
Comment (nlutsenko) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift]: I feel like these whitespace changes are unintentional. Do you think we could revert them?

**Comment by nlutsenko** [comment]
Requesting last time, simply because of white space changes.

File: Sources/ParseLiveQuery/Internal/ClientPrivate.swift (hunk 10 above)
Comment (noobs2ninjas) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift]: Am I reading this wrong? Im seeing that a whitespace was removed and then added back? I dont see why this is a problem.

Comment (nlutsenko) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift]: Looks like this is an unintentional whitespace change, probably added by Xcode?
Was added in commit 06222abfdf65c33d5fab6090b14c0a45966f611f and looks like never removed.

What I am referencing is all those whitespace only lines that didn't have whitespace and now they do.

Comment (noobs2ninjas) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift]: I just checked your current source to my current source and on line 212 there is whitespace....on both.

Comment (noobs2ninjas) [inline Sources/ParseLiveQuery/Internal/ClientPrivate.swift]: to be honest im not even sure how to revert that specific change.

**Comment by nlutsenko** [comment]
Ok, this is still whitespace full. :(
I am using https://github.com/ParsePlatform/ParseLiveQuery-iOS-OSX/pull/22/files as a reference - just take a look at where there is whitespace and you can remove it line by line.

**Comment by noobs2ninjas** [comment]
Ok got it. My god! Not sure if there is a default that XCode has that automatically adds those things but I swear 90% of it was never touched by myself. If there is a default pref that does that I think its a pretty stupid one personally.

**Comment by noobs2ninjas** [comment]
Sorry I didnt understand earlier. When you mean whitespace changes...you meant tab insertions. Im sure theres a more official term for it. Sorry about the miscommunication.

**Comment by nlutsenko** [comment]
All good. Thanks for fixing all of them. There is a tiny bit more, but at this point we'll just fix it as we go. Will merge as soon as it compiles.

**Comment by noobs2ninjas** [comment]
Ok. I found the setting in XCode that was causing the issue. Not sure why "include whitespace-only lines" isnt set by default". You'd think with them being so centered around version control in xcode that they would have it set by default.
